Make plugin info threadsafe

tylerkaraszewski opened the pull request:
@quinthar 
cc @cead22 @flodnv 

This addresses @quinthar's comment here:
https://github.com/Expensify/Bedrock/pull/41#discussion_r84563244

While I think the chances that anyone would implement pluginInfo in a way that would be likely to cause problems here are low, the existing code *was* non-threadsafe, so this fixes that.

I debated about using our traditional `void*` mutexes with `pthreads` for consistency's sake, but I think the new C++11 `mutex`es are so much nicer and more straightforward to work with, I decided this would be a good place to introduce them. I'd be up for replacing all of our existing mutex code with the stuff that's not in `<mutex>`.

quinthar commented:
Why not use the existing `SSynchronized` class that is already in wide use throughout the codebase, which is designed precisely for this?

tylerkaraszewski commented:
@quinthar: updated.

So the reason I wanted to use `std::mutex` at all is because I think it's much simpler and more straightforward to use than our existing mutex code. I also get that consistency across the codebase is nice to have as well.

So, I changed this code to use `SSynchronized`, but I swapped out the implementation of `SSynchronized` to use `std::mutex`. You can see this lets us get rid of a bunch of custom code, as well as skip various cleanup to deallocate mutexes, and we can declare mutexes as an actual `mutex` type, isntead of `void*` which I think is much more readable.

I hope this is the best of both worlds.
